include/nd_memory.h

```diff
@@ -0,0 +1,122 @@
+template <typename F>
+inline void for_each_contiguous_chunk(const nd_copy_layout& layout, F&& f) {
+	if(layout.contiguous_size == 0) return;
+
+	// nd_copy_layout is defined such that we can ignore num_complex_strides in this loop and will perform exactly one iteration for every non-complex stride.
+	size_t source_offset_0 = layout.offset_in_source;
+	size_t dest_offset_0 = layout.offset_in_dest;
+	for(size_t i = 0; i < layout.strides[0].count; ++i) {
+		size_t source_offset_1 = source_offset_0;
+		size_t dest_offset_1 = dest_offset_0;
+		for(size_t j = 0; j < layout.strides[1].count; ++j) {
+			f(source_offset_1, dest_offset_1, layout.contiguous_size);
+			source_offset_1 += layout.strides[1].source_stride;
+			dest_offset_1 += layout.strides[1].dest_stride;
+		}
+		source_offset_0 += layout.strides[0].source_stride;
+		dest_offset_0 += layout.strides[0].dest_stride;
+	}
+}
+
+/// From allocation `source_base` sized `source_range` starting at `offset_in_source`, to allocation `dest_base` sized `dest_range` starting at to
+/// `offset_in_dest`, copy `copy_range` elements of `elem_size` bytes.
+inline void nd_copy_host(const void* const source_base, void* const dest_base, const range<3>& source_range, const range<3>& dest_range,
+    const id<3>& offset_in_source, const id<3>& offset_in_dest, const range<3>& copy_range, const size_t elem_size) //
+{
+	const auto layout = layout_nd_copy(source_range, dest_range, offset_in_source, offset_in_dest, copy_range, elem_size);
+	for_each_contiguous_chunk(layout, [&](const size_t chunk_offset_in_source, const size_t chunk_offset_in_dest, const size_t chunk_size) {
+		memcpy(static_cast<std::byte*>(dest_base) + chunk_offset_in_dest, static_cast<const std::byte*>(source_base) + chunk_offset_in_source, chunk_size);
+	});
+}
+
+/// From allocation `source_base` spanning `source_box` to allocation `dest_base` spanning `dest_box`, copy `copy_box` elements of `elem_size` bytes.
+inline void nd_copy_host(
+    const void* const source_base, void* const dest_base, const box<3>& source_box, const box<3>& dest_box, const box<3>& copy_box, const size_t elem_size) //
+{
+	assert(source_box.covers(copy_box));
+	assert(dest_box.covers(copy_box));
+	nd_copy_host(source_base, dest_base, source_box.get_range(), dest_box.get_range(), copy_box.get_offset() - source_box.get_offset(),
+	    copy_box.get_offset() - dest_box.get_offset(), copy_box.get_range(), elem_size);
+}
+
+/// From allocation `source_base` spanning `source_box` to allocation `dest_base` spanning `dest_box`, copy `copy_region` elements of `elem_size` bytes.
+inline void nd_copy_host(const void* const source_base, void* const dest_base, const box<3>& source_box, const box<3>& dest_box, const region<3>& copy_region,
+    const size_t elem_size) //
+{
+	for(const auto& copy_box : copy_region.get_boxes()) {
+		nd_copy_host(source_base, dest_base, source_box, dest_box, copy_box, elem_size);
+	}
+}
```

Comment:
All of these interface have a user-supplied `elem_size`, rather than being templated on an element type. I guess this feels somewhat natural since they are low-level building blocks.

However, you can always make a size from a type but not the other way around, so if there were any underlying API we ever want to target with these that might benefit from type specialization we lose that option with this interface.

Reply:
In principle you're right, the most immediate effect this could have is missed optimization due to the lack of alignment information. However all current call sites are in the backend and are thus type-erased already, so this interface would need to continue existing even if we added a templated version.